datafusion/core/src/physical_optimizer/limit_pushdown.rs

```diff
@@ -0,0 +1,564 @@
+use datafusion_physical_plan::sorts::sort_preserving_merge::SortPreservingMergeExec;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+/// This rule inspects [`ExecutionPlan`]'s and pushes down the fetch limit from
+/// the parent to the child if applicable.
+#[derive(Default)]
+pub struct LimitPushdown {}
+
+impl LimitPushdown {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl PhysicalOptimizerRule for LimitPushdown {
+    fn optimize(
+        &self,
+        plan: Arc<dyn ExecutionPlan>,
+        _config: &ConfigOptions,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        plan.transform_down(push_down_limits).data()
+    }
+
+    fn name(&self) -> &str {
+        "LimitPushdown"
+    }
+
+    fn schema_check(&self) -> bool {
+        true
+    }
+}
+
+/// Helper enum to make skip and fetch easier to handle
+#[derive(Debug)]
+enum LimitExec {
+    Global(GlobalLimitExec),
+    Local(LocalLimitExec),
+}
+
+impl LimitExec {
+    fn input(&self) -> &Arc<dyn ExecutionPlan> {
+        match self {
+            Self::Global(global) => global.input(),
+            Self::Local(local) => local.input(),
+        }
+    }
+
+    fn fetch(&self) -> Option<usize> {
+        match self {
+            Self::Global(global) => global.fetch(),
+            Self::Local(local) => Some(local.fetch()),
+        }
+    }
+
+    fn skip(&self) -> usize {
+        match self {
+            Self::Global(global) => global.skip(),
+            Self::Local(_) => 0,
+        }
+    }
+
+    fn with_child(&self, child: Arc<dyn ExecutionPlan>) -> Self {
+        match self {
+            Self::Global(global) => {
+                Self::Global(GlobalLimitExec::new(child, global.skip(), global.fetch()))
+            }
+            Self::Local(local) => Self::Local(LocalLimitExec::new(child, local.fetch())),
+        }
+    }
+
+    fn with_fetch(&self, fetch: usize) -> Self {
+        match self {
+            Self::Global(global) => Self::Global(GlobalLimitExec::new(
+                global.input().clone(),
+                global.skip(),
+                Some(fetch),
+            )),
+            Self::Local(local) => {
+                Self::Local(LocalLimitExec::new(local.input().clone(), fetch))
+            }
+        }
+    }
+}
+
+impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+    fn from(limit_exec: LimitExec) -> Self {
+        match limit_exec {
+            LimitExec::Global(global) => Arc::new(global),
+            LimitExec::Local(local) => Arc::new(local),
+        }
+    }
+}
+
+/// Pushes down the limit through the plan.
+pub fn push_down_limits(
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+    let maybe_modified = if let Some(limit_exec) = extract_limit(plan.clone()) {
+        let child = limit_exec.input();
+        if let Some(child_limit) = child.as_any().downcast_ref::<LocalLimitExec>() {
+            let merged = try_merge_limits(&limit_exec, child_limit)?;
```

Comment:
What do we assume here for the placement of Limit operators? I guess we assume we can handle GlobalLimit -> LocalLimit and LocalLimit -> LocalLimit. Can only these conditions exist and others cannot? If so, a comment would be good.

Reply:
I don't have a deep enough understanding but it seems all sql logic tests have a single `GlobalLimitExec` near the top and one or multiple `LocalLimitExec`s below. Therefore it seems we only have GlobalLimit -> LocalLimit and LocalLimit -> LocalLimit cases but I'm not sure if others are possible

Reply:
You are right; the following examples are not appearing in the plans, but they are theoretically possible: LocalLimitExec can be on top of GlobalLimitExec (LocalLimit -> GlobalLimit), and two consecutive GlobalLimitExecs (GlobalLimit -> GlobalLimit). Since we know that these operators work on a single partition for sure, we could also merge these into a single GlobalLimit after a slight refactor of your code.